rfd/0083-machine-id-host-certs.md

```diff
@@ -0,0 +1,150 @@
+
+   For example, as implemented today, `equals(host_cert.principals, "foo")`
+   already iterates over all LHS arguments and ANDs the result. A hypothetical
+   `ends_with()` should behave the same way.
+
+3. Evaluating `where` clauses in [`GenerateHostCert()`]. These follow an
+   entirely different codepath than regular Teleport resource access
+   ([`CheckAccessToRule()`]), so we'll likely need to implement special handling
+   for this special non-resource, since we don't have a single name / namespace /
+   etc.
+
+   [Session Access] has a custom predicate implementation that would mostly suit
+   our needs, though we'll need to follow standard RBAC rule behavior.
+
+[`GenerateHostCert()`]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/auth/auth_with_roles.go#L2139
+[`CheckAccessToRule()`]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/services/role.go#L2309
+[Session Access]: https://github.com/gravitational/teleport/blob/82c520c8183553f310459c3b4a96b70065ee268a/lib/auth/session_access.go#L126
+
+### UX
+
+We have several options for issuing host certificates. Factors to consider
+include:
+1. Certificate lifespan. Do these certificates need to expire every hour?
+2. Does issuing them every hour create additional UX problems of its own?
+3. Does `sshd` gracefully reload these certificates, and if so, how can users be
+   expected to gracefully reload `sshd` when `tbot` is running as a user
+   process?
+
+Per (3), we have an open issue to help notify applications of changed
+certificates (https://github.com/gravitational/teleport/issues/11264). It's
+unclear if OpenSSH gracefully handles certificate rotations, though, and
+disconnecting sessions every 20 minutes would not be ideal.
+
+#### Option 0: No-op
+
+Users can already use bot identities to generate host certs with the `host_cert`
+permission granted, by passing the Machine ID `identity` file to `tctl`:
+
+```bash
+$ tctl -i path/to/identity auth sign \
+    --host=foo.example.com \
+    --format=openssh \
+    --out=myhost
+```
+
+This UX works fine in tandem with secure certificate issuance, but users are
+left to issue certs manually and to inform `sshd` of the change. If managed by
+cron / systemd timers / Ansible / etc, this could arguably be more 
+straightforward than `tbot` file change notifications (but complicates handling
+of CA rotation events).
+
+#### Option 1: Config template
+
+Config Templates in Machine ID are refreshed each iteration of the bots usual
+renewal loop (20min by default). This would be the most straightforward
+integration approach and we'd approach it like we do other special cert types
+for apps / databases / k8s clusters.
+
+Users would request a host cert via the config file:
+
+```yaml
+destinations:
+  - directory: /opt/machine-id
+    configs:
+      - host_cert:
+          principals: [foo.example.com]
```

Comment:
That's a good question, I think we could do that (or provide it via some special word). It does run into the same problems though - malicious hosts could request certs to impersonate another server. I bet as with https://github.com/gravitational/teleport/issues/13483 we could also grant a default allow exception for the client's own IP?

Otherwise, by default, the `host_cert` rule doesn't have a `where` clause so bots with the rule can issue whichever certs they'd like. If users want to lock this down they'll need some predictable pattern to the principals, whether they're IPs or DNS names. We can use `starts_with()` or `matches()` rules (or a combination) to match against all possibilities.

Reply:
Yeah, there's some work we need to do on #15521 that will let the auth service see the client's real address, then we could at least have a special keyword or similar to allow this.